### ISSUE TITLE
Is it the same adding weight decay to all the layers (including input and output layer) than adding the weight decay term to the cost function?

### ISSUE THREAD

**Issue by EnriqueSMarquez** [fchollet/keras#2717]
Hello,

Using other frameworks I have used weight decay on the cost function rather than layer wise. How does weight decay per layer work?

And if I wanted to do weight decay to the cost function how would it be on Keras. Thank you.

Cheers,
EM

**Comment by joelthchao** [comment]
Weight decay is added by `Regularizers` ([docs](http://keras.io/regularizers/)). It only works on layers you add this argument.

**Comment by michaelschleiss** [comment]
Can somebody explain the original question? I also know weight decay only in the cost function. So how should I impelement this in Keras?

**Comment by sebastianbk** [comment]
@michaelschleiss, you can take a look at the [docs](https://keras.io/regularizers/) that @joelthchao referred to.

The question is about whether or not it is possible to add the weight decay term manually to the global cost function in a Keras model. While it is possible to create your own cost function, in addition to Keras' preimplemented cost functions ([see the docs here](https://keras.io/losses/), it doesn't seems as if there is an easy to get hold of the weight matrix, which you would need in order to add the weight decay term to the cost. Only the true labels and the predicted labels seem to be accessible from within a custom cost method.

If you are interested in adding regularization to your network, your best bet would be to consider layer-specific regularizers that you would have to add one by one.

**Comment by perryshao** [comment]
I have also this problem. I think the global weight decay is actually not equal to the layer-wise weight decay.

**Comment by joelthchao** [comment]
@perryshao global weight decay is just summation of all layer-wise weight decays. Therefore, if you want to have a "global weight decay" in your loss function, just put Regularizer on all the layers with trainable weights.

**Comment by zekun-li** [comment]
@joelthchao If I use Regularizer on all the layers, what the regularization parameter should be? Suppose I have global weight decay rate to be 0.005 in caffe, is it the same as using 0.005 for each layer in keras?

**Comment by joelthchao** [comment]
@zekun-li Yes.

**Comment by emoen** [comment]
with deeper and deeper layers - it would be nice to have a cost function weight decay so you dont have to add it to each layer.

**Comment by tutysara** [comment]
This could be of help to anyone who is looking for how to code this, ty.

``` python
for layer in my_model.layers:
    if hasattr(layer, 'kernel_regularizer'):
        layer.kernel_regularizer= regularizers.l2(weight_decay)
```

**Comment by HerocypheR** [comment]
@tutysara thanks for example. is this same as : 

```
trainoptmodel = Sequential()
trainoptmodel.add(Dense(25,input_dim=10, activation='tanh', kernel_regularizer=regularizers.l2(0.01)))
trainoptmodel.add(Dense(3,activation="linear", kernel_regularizer=regularizers.l2(0.01))

``` 
And mean of this code is global weight decay is identified as 0.01 ?

**Comment by mjmikulski** [comment]
@tutysara this does **not** work. 

I just lost an hour to investigate it. Adding regularizers this way will not affect `layer.losses`. Layers losses are constructed when adding layer, so adding regularizers after layer was constructed will not affect model loss.

**Comment by tutysara** [comment]
@mjmikulski Hey sorry, didn't know it doesn't take effect after layer construction. I based my implementation based on @joelthchao comment. How did you end up solving it?

**Comment by perryshao** [comment]
@joelthchao In the view of mathematical equation of L2 norm, it is obvious that the global weight decay using L2 norm is not exactly same with the layer-wise weight decay using L2 norm. For example, sqrt(x1^2+x2^2)+sqrt(y1^2+y2^2), and sqrt(x1^2+x2^2+y1^2+y2^2), suppose that  x vector is the weights of layer 1 and y is the weights of layer 2.

**Comment by mjmikulski** [comment]
@perryshao From mathematical point of view you are perfectly right - there is a sqrt in L2. 

But in keras (or maybe more generally in ML), L2 means rather "sum of squares", not really a proper L2 norm, see keras implementation:
```python
class L1L2(Regularizer):
    # ...
    def __call__(self, x):
        regularization = 0.
        if self.l1:
            regularization += K.sum(self.l1 * K.abs(x))
        if self.l2:
            regularization += K.sum(self.l2 * K.square(x))
        return regularization
```

**Comment by perryshao** [comment]
@mjmikulski Yes, you are right. I checked with the codes, where it is not really a proper L2 norm, just "a sum of squares". thank you for your reply.

**Comment by apatsekin** [comment]
@EnriqueSMarquez
>>weight decay is exactly the same what L2 norm
Apparently it's not. Here is an explanation why:
https://bbabenko.github.io/weight-decay/
TL/DR: Weight Decay is subtracted directly from weights on each step as is, but L2_reg is added to loss, hence it affects weights as derivative (multiplied by 2). To be consistent with weight decay it should be divided by 2 in loss function.
When using sophisticated optimizers there are some other differences:
http://www.fast.ai/2018/07/02/adam-weight-decay/

The problem of adding kernel_regularizer to the existing layer is still unsolved for me. Even though there is a stackoverflow question claims it should work:
https://stackoverflow.com/questions/48330137/adding-regularizer-to-an-existing-layer-of-a-trained-model-without-resetting-wei

**Comment by mjmikulski** [comment]
@apatsekin good point! I have not thought about it.

Let me paste here part of bbabenko [article](https://bbabenko.github.io/weight-decay/) you mention:

> In frameworks that implement L2 regularization, the gradients the solver is using are for “total loss”, which has L2 regularization baked into it (this fact is typically abstracted away from the solver — it doesn’t “know” anything about whether a regularization term was included in the loss or not). In frameworks that implement weight decay, the solver is considering only the gradient for  L.

What I said:

> weight decay is exactly the same what L2 norm

is true for a plain gradient descend (up to factor 2 which is irrelevant when you do a hyperparameter search, but is sth to remember when you try to reproduce an experiment). Situation gets more complicated when you accumulate the gradient some way. Thanks @apatsekin for pointing it out.


Anyway, what is wrong with [this stackoverflow question](https://stackoverflow.com/questions/48330137/adding-regularizer-to-an-existing-layer-of-a-trained-model-without-resetting-wei)? The answer makes a point and explains OP's observation.

**Comment by apatsekin** [comment]
> Anyway, what is wrong with this stackoverflow question? The answer makes a point and explains OP's observation.

As you pointed earlier in this thread:

> I just lost an hour to investigate it. Adding regularizers this way will not affect layer.losses

_Recompile_ doesn't pick it up.  The stackoverflow thread claims the opposite.

**Comment by batzner** [comment]
```
# Add weight decay to the whole model
regularizer = tf.keras.regularizers.l2(WEIGHT_DECAY)
decay_attributes = ['kernel_regularizer', 'bias_regularizer',
                    'beta_regularizer', 'gamma_regularizer']
for layer in model.layers:
    for attr in decay_attributes:
        if hasattr(layer, attr):
            setattr(layer, attr, regularizer)
```

**Comment by nicolefinnie** [comment]
Thanks @batzner, very well written. It did change the loaded model's regularizers. 

> ```python
> # Add weight decay to the whole model
> regularizer = tf.keras.regularizers.l2(WEIGHT_DECAY)
> decay_attributes = ['kernel_regularizer', 'bias_regularizer',
>                     'beta_regularizer', 'gamma_regularizer']
> 
> for layer in model.layers:
>     for attr in decay_attributes:
>         if hasattr(layer, attr):
>             setattr(layer, attr, regularizer)
> ```

For example, the kernel regularizer of the first conv layer of Keras `mobilenet_v2` with the weight decay = `0.00004` looks as follows:

```
Conv1: {'class_name': 'L1L2', 'config': {'l1': 0.0, 'l2': 3.9999998989515007e-05}} 
```

**Comment by mrluin** [comment]
> e this, ty.

ni

> Thanks @batzner, very well written. It did change the loaded model's regularizers.
> 
> > ```python
> > # Add weight decay to the whole model
> > regularizer = tf.keras.regularizers.l2(WEIGHT_DECAY)
> > decay_attributes = ['kernel_regularizer', 'bias_regularizer',
> >                     'beta_regularizer', 'gamma_regularizer']
> > 
> > for layer in model.layers:
> >     for attr in decay_attributes:
> >         if hasattr(layer, attr):
> >             setattr(layer, attr, regularizer)
> > ```
> 
> For example, the kernel regularizer of the first conv layer of Keras `mobilenet_v2` with the weight decay = `0.00004` looks as follows:
> 
> ```
> Conv1: {'class_name': 'L1L2', 'config': {'l1': 0.0, 'l2': 3.9999998989515007e-05}} 
> ```

Nice solution, thank you~

**Comment by mrluin** [comment]
> Thanks @batzner, very well written. It did change the loaded model's regularizers.
> 
> > ```python
> > # Add weight decay to the whole model
> > regularizer = tf.keras.regularizers.l2(WEIGHT_DECAY)
> > decay_attributes = ['kernel_regularizer', 'bias_regularizer',
> >                     'beta_regularizer', 'gamma_regularizer']
> > 
> > for layer in model.layers:
> >     for attr in decay_attributes:
> >         if hasattr(layer, attr):
> >             setattr(layer, attr, regularizer)
> > ```
> 
> For example, the kernel regularizer of the first conv layer of Keras `mobilenet_v2` with the weight decay = `0.00004` looks as follows:
> 
> ```
> Conv1: {'class_name': 'L1L2', 'config': {'l1': 0.0, 'l2': 3.9999998989515007e-05}} 
> ```

Nice solution, thank you

**Comment by lars76** [comment]
@batzner code is not working, because it only changes the model config, but does not create the addition tensors itself. One can check it easily by running 

```python
for l in model.layers:
	print(l.losses)
```

A workaround is the following code

```python
def create_model():
    model = your_model()
    model.save_weights("tmp.h5")

    # optionally do some other modifications (freezing layers, adding convolutions etc.)
    ....

    regularizer = l2(WEIGHT_DECAY / 2)
    for layer in model.layers:
        for attr in ['kernel_regularizer', 'bias_regularizer']:
            if hasattr(layer, attr) and layer.trainable:
                setattr(layer, attr, regularizer)

    out = model_from_json(model.to_json())
    out.load_weights("tmp.h5", by_name=True)

    return out
```

Weight decay refers only to the bias and kernel weights (not BN see https://arxiv.org/pdf/1810.12281.pdf). This means for SGD use 'kernel_regularizer' + 'bias_regularizer' with l2(WEIGHT_DECAY / 2). And for Adam one has to create a custom optimizer based on tf.contrib.opt.AdamWOptimizer (see https://arxiv.org/pdf/1711.05101.pdf). As somebody else already noted, when in a paper, it is written weight decay = 0.0005, we have to divide by 2 (due to the derivative of ||weight||_2^2).

**Comment by nicolefinnie** [comment]
> @batzner code is not working, because it only changes the model config, but does not create the addition tensors itself. One can check it easily by running
> 
> ```python
> for l in model.layers:
> 	print(l.losses)
> ```
> 
> A workaround is the following code
> 
> ```python
> def create_model():
>     model = your_model()
>     model.save_weights("tmp.h5")
> 
>     # optionally do some other modifications (freezing layers, adding convolutions etc.)
>     ....
> 
>     regularizer = l2(WEIGHT_DECAY / 2)
>     for layer in model.layers:
>         for attr in ['kernel_regularizer', 'bias_regularizer']:
>             if hasattr(layer, attr) and layer.trainable:
>                 setattr(layer, attr, regularizer)
> 
>     out = model_from_json(model.to_json())
>     out.load_weights("tmp.h5", by_name=True)
> 
>     return out
> ```
> 
> Weight decay refers only to the bias and kernel weights (not BN see https://arxiv.org/pdf/1810.12281.pdf). This means for SGD use 'kernel_regularizer' + 'bias_regularizer' with l2(WEIGHT_DECAY / 2). And for Adam one has to create a custom optimizer based on tf.contrib.opt.AdamWOptimizer (see https://arxiv.org/pdf/1711.05101.pdf). As somebody else already noted, when in a paper, it is written weight decay = 0.0005, we have to divide by 2 (due to the derivative of ||weight||_2^2).

@lars76 Thanks for pointing it out! I didn't know only the config got changed and this [issue 12053](https://github.com/keras-team/keras/issues/12053) was just opened 6 days ago. Also,the way  keras handles the regularization  loss differs from tensorflow, see [issue 21587](https://github.com/tensorflow/tensorflow/issues/21587).  However, if we load an empty model (such as predefined model in Keras) without loading any pre-trained weights and losses, modify the regularizers of the model, and compile the model, would this still be a problem? The `layer.losses` are empty before the training anyway. Thanks!

**Comment by batzner** [comment]
@lars76 thank you for the hint! I updated my comment to make clear that the naïve solution only changes the model's config. I also added a note to highlight that L2 regularization doesn't directly correspond to weight decay.

As for the paper you referenced (https://arxiv.org/pdf/1810.12281.pdf), it states:

> For clarity, we ignore the parameters γ and β, which do not impact the performance in practice. 

As far as I understand it, they disable the learnable affine parameters of Batch Normalization completely and therefore don't regularize them. But if I were to enable them, is it harmful to also regularize them?

**Comment by lars76** [comment]
No, you can also regularize them. I just disabled gamma/beta, because the authors of the paper wrote that it made no difference performance-wise and older papers include only regular weights/biases (since BN didn't exist). Maybe you're even right, it might be more logical to set regularizers to both weights too, since beta corresponds to bias weights and gamma to regular weights. I think then the correct definition for SGD weight decay is "Loss function + 1/2 sum w_i^2 where w_i are all trainable weights".

**Comment by mathmanu** [comment]
How about this:

	# a utility function to add weight decay after the model is defined.
	def add_weight_decay(model, weight_decay):
		if (weight_decay is None) or (weight_decay == 0.0):
			return

		# recursion inside the model
		def add_decay_loss(m, factor):
			if isinstance(m, tf.keras.Model):
				for layer in m.layers:
					add_decay_loss(layer, factor)
			else:
				for param in m.trainable_weights:
					with tf.keras.backend.name_scope('weight_regularizer'):
						regularizer = lambda: tf.keras.regularizers.l2(factor)(param)
						m.add_loss(regularizer)

		# weight decay and l2 regularization differs by a factor of 2
		add_decay_loss(model, weight_decay/2.0)
		return

**Comment by philipperemy** [comment]
If your model contains nested Models object, you can use this function:

```python
def add_l2_weight_decay(net: Model, weights_decay=5e-4):
    reg = l2(weights_decay)
    for layer in net.layers:
        if isinstance(layer, Model):
            add_l2_weight_decay(layer, weights_decay)
        for attr in ['kernel_regularizer', 'bias_regularizer']:
            if hasattr(layer, attr) and layer.trainable:
                setattr(layer, attr, reg)

```

**Comment by nixingyang** [comment]
Here is the snippet that I come up with. It is very generalized and should work in all cases. Previous answers do not work correctly if the "trainable" properties have been modified.

```python
import numpy as np
from tensorflow.keras.layers import BatchNormalization, Dense
from tensorflow.keras.regularizers import l2
from tensorflow.python.keras.layers.convolutional import Conv


def gather_items(model):
    # Get the model and all its children
    items = model._gather_layers()  # pylint: disable=protected-access

    # Separate models and layers
    models, layers = [], []
    for item in items:
        # A model should have submodules
        if len(item.submodules) > 0:
            models.append(item)
        else:
            layers.append(item)

    return models, layers


def specify_regularizers(model,
                         kernel_regularization_factor=0,
                         bias_regularization_factor=0,
                         gamma_regularization_factor=0,
                         beta_regularization_factor=0):

    def _assign_regularizer(layer, attribute_name, switch_name,
                            regularization_factor, summary):
        # regularization factor should be a positive number
        # layer should have the attribute
        if regularization_factor <= 0 or not hasattr(layer, attribute_name):
            return

        # Disable the additional check if switch_name is None
        # (only relevant to the kernel parameter in Conv and Dense layers)
        # Skip if the switch_name attribute is False
        # (the parameter of interest is not enabled)
        if switch_name is not None and not getattr(layer, switch_name):
            return

        # Set the regularizer in layer
        # (only relevant if one needs to save configurations of the model)
        regularizer = l2(l=regularization_factor)
        setattr(layer, attribute_name, regularizer)

        # Add the regularization loss term
        # https://github.com/tensorflow/tensorflow/blob/v2.2.2/tensorflow/python/keras/engine/base_layer.py#L578-L585
        variable = getattr(layer, attribute_name.split("_")[0])
        name_in_scope = variable.name[:variable.name.find(':')]
        layer._handle_weight_regularization(  # pylint: disable=protected-access
            name_in_scope, variable, regularizer)

        # Update summary
        key = "{}:{}".format(layer.__class__.__name__, attribute_name)
        if key not in summary:
            summary[key] = 0
        summary[key] += 1

    # Assign regularizers
    summary = {}
    losses_num = len(model.losses)
    print("Specifying regularizers for {} ...".format(model.name))
    _, layers = gather_items(model)
    for layer in layers:
        if isinstance(layer, (Conv, Dense)):
            _assign_regularizer(
                layer=layer,
                attribute_name="kernel_regularizer",
                switch_name=None,
                regularization_factor=kernel_regularization_factor,
                summary=summary)
            _assign_regularizer(
                layer=layer,
                attribute_name="bias_regularizer",
                switch_name="use_bias",
                regularization_factor=bias_regularization_factor,
                summary=summary)
        elif isinstance(layer, BatchNormalization):
            _assign_regularizer(
                layer=layer,
                attribute_name="gamma_regularizer",
                switch_name="scale",
                regularization_factor=gamma_regularization_factor,
                summary=summary)
            _assign_regularizer(
                layer=layer,
                attribute_name="beta_regularizer",
                switch_name="center",
                regularization_factor=beta_regularization_factor,
                summary=summary)
    print(summary)

    # Sanity check
    if summary:
        added_losses_num = np.sum(list(summary.values()))
        print("Added {} regularization terms.".format(added_losses_num))
        assert len(model.losses) == losses_num + added_losses_num
```
